### PR TITLE
feat(universe): add ker and kama to flux

### PR DESCRIPTION
**Pull request by ewendai**
### Done checklist
- [x] docs/SPEC.md updated
- [x] Test cases written

#1541 
#1542

### CODE REVIEW

File: stdlib/universe/universe.flux [hunk 1]
```diff
@@ -2,6 +2,9 @@ package universe
 
 import "system"
 import "date"
+import "math"
+import "strings"
+import "regexp"
```
Comment (wolffcm) [inline stdlib/universe/universe.flux]: I see you took these packages out of the prelude in `compile.go`, curious about why?

File: stdlib/universe/kaufmansAMA_test.go [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+	"github.com/influxdata/flux/stdlib/universe"
+	"testing"
+)
+
+func TestKamaOperation_Marshaling(t *testing.T) {
+	data := []byte(`{"id":"kaufmansAMA","kind":"kaufmansAMA","spec":{"n":1}}`)
+	op := &flux.Operation{
+		ID: "kaufmansAMA",
+		Spec: &universe.KamaOpSpec{
+			N: 1,
```
Comment (wolffcm) [inline stdlib/universe/kaufmansAMA_test.go]: You could verify that the `columns` parameter is marshaled correctly here too.

File: stdlib/universe/kaufmansAMA_test.go [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+	}
+	querytest.OperationMarshalingTestHelper(t, data, op)
+}
+
+func TestKama_PassThrough(t *testing.T) {
+	executetest.TransformationPassThroughTestHelper(t, func(d execute.Dataset, c execute.TableBuilderCache) execute.Transformation {
+		s := universe.NewkamaTransformation(
+			d,
+			c,
+			&universe.KamaProcedureSpec{},
```
Comment (wolffcm) [inline stdlib/universe/kaufmansAMA_test.go]: So I guess that kama will just forward its input tables to its output if `n` is zero and `columns` is empty... Do both of these conditions have to be true for kama to be a passthrough?

Do we have other functions that behave this way, i.e., they become pass through functions for certain parameter values.  I could imagine returning an error for cases like this as well.

File: stdlib/universe/kaufmansAMA_test.go [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+					{Label: "_time", Type: flux.TTime},
+					{Label: "_value", Type: flux.TFloat},
+				},
+				Data: [][]interface{}{
+					{execute.Time(14), 1.4444444444444446},
+					{execute.Time(14), 2.5802469135802473},
+				},
+			}},
+		},
+	}
```
Comment (wolffcm) [inline stdlib/universe/kaufmansAMA_test.go]: You might want to add test coverage for some of the error cases too (like if a column is not a type that kama will accept).

File: stdlib/universe/kaufmansAMA_test.go (hunk 3 above)
Comment (wolffcm) [reply stdlib/universe/kaufmansAMA_test.go]: Looking over code in the `Process` method, it seems like `columns:[]` should be a pass through, but there should be an error returned if `n` is not positive.

File: stdlib/universe/kaufmansAMA.go [hunk 5]
```diff
@@ -0,0 +1,259 @@
...
+						values = append(values, float64(arr.Value(i)))
+					}
+				case flux.TFloat:
+					arr := cr.Floats(j)
+					values = arr.Float64Values()
+				}
+				kers := t.makeKERS(values, int(t.n))
+				var kama float64
+				sc := 0.0
+				kama = values[t.n-1]
```
Comment (wolffcm) [inline stdlib/universe/kaufmansAMA.go]: if `t.n` is greater than the number of rows in the table, then this will `panic`, right?  What would the best behavior for this be?  It could either return an error, or just return an empty table.  What does EMA do?

File: stdlib/universe/kaufmansAMA_test.go [hunk 6]
```diff
@@ -0,0 +1,209 @@
...
+		data []flux.Table
+		want []*executetest.Table
+	}{
+		{
+			name: "flux.TInt input",
+			spec: &universe.KamaProcedureSpec{
+				N:       10,
+				Columns: []string{"_value"},
+			},
+			data: []flux.Table{&executetest.Table{
```
Comment (wolffcm) [inline stdlib/universe/kaufmansAMA_test.go]: Some of the tests for EMA use row-wise tables:
https://github.com/influxdata/flux/blob/master/stdlib/universe/exponential_moving_average_test.go#L78

Could you add at least one test case here that does as well?

Comment (wolffcm) [reply stdlib/universe/kaufmansAMA_test.go]: This will force the table block size to be exactly one row, so you can verify that KAMA works right for tables that span multiple blocks.

File: stdlib/universe/kaufmansAMA.go [hunk 7]
```diff
@@ -0,0 +1,259 @@
...
+				for i := int(t.n); i < cr.Len(); i++ {
+					if err := builder.AppendValue(j, execute.ValueForRow(cr, i, j)); err != nil {
+						return err
+					}
+				}
+			} else {
+				var values []float64
+				switch c.Type {
+				case flux.TInt:
+					arr := cr.Ints(j)
```
Comment (wolffcm) [inline stdlib/universe/kaufmansAMA.go]: I don't think this will work quite right for large tables.  Tables are partitioned into multiple "blocks" that by default can have up to 1000 rows.  The issue is that `table.Do` will be called multiple times for each block.  So at the 1001st row, it will start over assuming that it's a new table.

EMA uses data structures in a package called `moving_average`

Comment (wolffcm) [reply stdlib/universe/kaufmansAMA.go]: EMA uses data structures from the pacakge `flux/internal/moving_average` to make this work right.  Since this is also a moving average, I wonder if you could reuse this code?
https://github.com/influxdata/flux/blob/master/stdlib/universe/exponential_moving_average.go#L154
https://github.com/influxdata/flux/tree/master/internal/moving_average

File: docs/SPEC.md [hunk 8]
```diff
@@ -3550,7 +3550,107 @@ from(bucket: "telegraf/autogen"):
     |> relativeSearchIndex(n: 14)
 ```
 
+#### Kaufman's Efficiency Ratio
+
+Kaufman's Efficiency Ratio indicator is created by taking the absolute value of CMO and dividing by 100 to give the 0 to 1 range.
```
Comment (wolffcm) [inline docs/SPEC.md]: Can you expand CMO to its full name?  I think it will make this easier to understand.  Also, maybe add a sentence that tells what KER does, like `Kaufman's Efficiency Ratio provides a way of measuring the volatility of a series.`

File: docs/SPEC.md [hunk 9]
```diff
@@ -3550,7 +3550,107 @@ from(bucket: "telegraf/autogen"):
...
+
+The default column `KaufmansER` takes in is "_value".
+
+| Name        | Type     | Description
+| ----        | ----     | -----------
+| n           | int      | N specifies the period.
+
```
Comment (wolffcm) [inline docs/SPEC.md]: You want to include the `columns` parameter here, right?

Comment (wolffcm) [reply docs/SPEC.md]: Ah, there's no `columns` parameter for this one.

File: docs/SPEC.md [hunk 10]
```diff
@@ -3550,7 +3550,107 @@ from(bucket: "telegraf/autogen"):
...
+ Example of script:
+ ```
+ from(bucket: "telegraf/autogen"):
+     |> range(start: -7d)
+     |> kaufmansER(n: 10)
+ ```
+
+#### Kaufman's Adaptive Moving Average
+
+Kaufman's Adaptive Moving Average is created by taking the absolute value of CMO and dividing by 100 to give the 0 to 1 range.
```
Comment (wolffcm) [inline docs/SPEC.md]: Same comment here regarding CMO.

File: compile.go [hunk 11]
```diff
@@ -88,10 +88,6 @@ var (
 	prelude = []string{
 		"universe",
 		"influxdata/influxdb",
-		"math",
```
[the remark this answers is not in the document]
Comment (ewendai) [reply compile.go]: @aanthony1243 removed those because otherwise "math.abs" would not perform as expected in `universe.flux`. Has to do with "math" should be imported first but sometimes it would be imported after because maps are processed randomly.

File: stdlib/universe/universe.flux (hunk 1 above)
Comment (ewendai) [reply stdlib/universe/universe.flux]: re: above

File: stdlib/testing/testdata/kama_v2.flux [hunk 12]
```diff
@@ -0,0 +1,78 @@
...
+import "testing"
+
+option now = () => (2030-01-01T00:00:00Z)
+
+inData = "
+#datatype,string,long,dateTime:RFC3339,double
+#group,false,false,false,false
+#default,_result,,,
+,result,table,_time,_value
+,,0,2018-05-22T00:00:00Z,110.46
```
[the remark this answers is not in the document]
Comment (ewendai) [reply stdlib/testing/testdata/kama_v2.flux]: done

File: compile.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (aanthony1243) [reply compile.go]: there was a data race wrt the order the packages were loaded.   moving them to `universe.flux` has the same outcome

File: stdlib/universe/kaufmansAMA.go (hunk 5 above)
Comment (ewendai) [reply stdlib/universe/kaufmansAMA.go]: EMA doesn't consider this case.

File: compile.go [hunk 13]
```diff
@@ -429,9 +425,6 @@ func (t *TableObject) PolyType() semantic.PolyType {
 func (t *TableObject) Str() string {
 	panic(values.UnexpectedKind(semantic.Object, semantic.String))
 }
-func (t *TableObject) Bytes() []byte {
```
Comment (wolffcm) [inline compile.go]: What are these changes?  Seems unrelated to the KAMA work?

Comment (ewendai) [reply compile.go]: i think i broke it when rebasing. fixing rn

File: stdlib/universe/kaufmansAMA.go [hunk 14]
```diff
@@ -0,0 +1,313 @@
+package universe
+
+import (
+	"fmt"
+	"github.com/influxdata/flux/interpreter"
```
Comment (wolffcm) [inline stdlib/universe/kaufmansAMA.go]: This import should be below with the other ones.

File: stdlib/universe/kaufmansAMA.go [hunk 15]
```diff
@@ -0,0 +1,313 @@
...
+	t.sumDown = 0
+	t.diffNAgo = make([]float64, t.n)
+	t.count = 0
+
+	err := tbl.Do(func(cr flux.ColReader) error {
+		if cr.Len() == 0 || len(cr.Cols()) == 0 {
+			return nil
+		}
+
+		if cr.Len() < int(t.n) {
```
Comment (wolffcm) [inline stdlib/universe/kaufmansAMA.go]: Hmm, I don't think this is quite right.  The value returned by `cr.Len()` only reflects the number of rows in this particular chunk of the table.  But the entire table could be composed of many chunks.  So it's okay if `cr.Len()` is less than N.  Flux should get the same answer no matter what the chunk size is.

File: stdlib/universe/kaufmansAMA.go [hunk 16]
```diff
@@ -0,0 +1,301 @@
...
+			doKAMA[j] = true
+		} else {
+			_, err := builder.AddCol(c)
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	t.prevValue = 0
```
Comment (wolffcm) [inline stdlib/universe/kaufmansAMA.go]: Are these variables only used inside the call to `tbl.Do()` below?  If so, can you make them local variables to the Process method?

File: docs/SPEC.md [hunk 17]
```diff
@@ -3613,7 +3613,107 @@ from(bucket: "telegraf/autogen"):
...
+ ```
+
+#### Kaufman's Adaptive Moving Average
+
+Kaufman's Adaptive Moving Average is designed to account for market noise or volatility. This trend-following indicator can be used to identify the overall trend, time turning points and filter data movements.
+
+| Name        | Type     | Description
+| ----        | ----     | -----------
+| n           | int      | N specifies the period.
+| columns     | []string | Columns is the list of all columns that `KaufmansAMA` should be performed on and is assumed to be ["_value"] if left unspecified.
```
Comment (wolffcm) [inline docs/SPEC.md]: I guess this needs to just be `column`.

File: stdlib/universe/kaufmansAMA_test.go [hunk 18]
```diff
@@ -0,0 +1,309 @@
...
+			name: "flux.TInt input rowWise",
+			spec: &universe.KamaProcedureSpec{
+				N:      10,
+				Column: "_value",
+			},
+			data: []flux.Table{&executetest.RowWiseTable{
+				Table: &executetest.Table{
+					ColMeta: []flux.ColMeta{
+						{Label: "_value", Type: flux.TInt},
+					},
```
Comment (wolffcm) [inline stdlib/universe/kaufmansAMA_test.go]: Can you add a test case that uses rowwise table *and* also has columns that besides just `_value`, like maybe a `_time` column?

File: stdlib/universe/kaufmansAMA.go [hunk 19]
```diff
@@ -0,0 +1,301 @@
...
+	t.count = 0
+
+	err := tbl.Do(func(cr flux.ColReader) error {
+		if cr.Len() == 0 || len(cr.Cols()) == 0 {
+			return nil
+		}
+
+		for j, c := range cr.Cols() {
+			if !doKAMA[j] {
+				for i := int(t.n); i < cr.Len(); i++ {
```
Comment (wolffcm) [inline stdlib/universe/kaufmansAMA.go]: I'm not sure this will work quite right if there is more than one chunk in the table...  The first chunk will only have some rows copied to the output, but the latter chunks would need to have all their values copied.  You can find out if your on the first chunk by lookng at `t.count` I guess.
